python/ray/train/jax/jax_trainer.py

```diff
@@ -0,0 +1,205 @@
+            it will be fit on the training dataset. All datasets will be transformed
+            by the ``preprocessor`` if one is provided.
+        preprocessor: A ``ray.data.Preprocessor`` to preprocess the
+            provided datasets.
+        resume_from_checkpoint: A checkpoint to resume training from.
+    """
+
+    def __init__(
+        self,
+        train_loop_per_worker: Union[Callable[[], None], Callable[[Dict], None]],
+        *,
+        train_loop_config: Optional[Dict] = None,
+        jax_config: Optional[JaxConfig] = None,
+        scaling_config: Optional[ScalingConfig] = None,
+        dataset_config: Optional[Dict[str, DatasetConfig]] = None,
+        run_config: Optional[RunConfig] = None,
+        datasets: Optional[Dict[str, GenDataset]] = None,
+        preprocessor: Optional["Preprocessor"] = None,
+        resume_from_checkpoint: Optional[Checkpoint] = None,
+    ):
+        if not jax_config:
+            jax_config = JaxConfig()
+
+        super(JaxTrainer, self).__init__(
+            train_loop_per_worker=train_loop_per_worker,
+            train_loop_config=train_loop_config,
+            backend_config=jax_config,
+            scaling_config=scaling_config,
+            dataset_config=dataset_config,
+            run_config=run_config,
+            datasets=datasets,
+            preprocessor=preprocessor,
+            resume_from_checkpoint=resume_from_checkpoint,
+        )
+
+    @classmethod
+    def _validate_scaling_config(cls, scaling_config: ScalingConfig) -> ScalingConfig:
+        """Return scaling config dataclass after validating updated keys."""
+        ensure_only_allowed_dataclass_keys_updated(
+            dataclass=scaling_config,
+            allowed_keys=cls._scaling_config_allowed_keys,
+        )
+
+        use_gpu = scaling_config.use_gpu
+        use_tpu = (
+            scaling_config.resources_per_worker
+            and scaling_config.resources_per_worker.get("TPU", 0) > 0
+        )
+
+        # cpu parallelism is not supported in jax
+        if not (use_gpu or use_tpu):
+            logger.warning(
+                "CPU parallelism is not supported in jax. "
+                "Please use distributed GPU or TPU training instead. "
+                "Currently, the code is still running on CPU, "
+                "but there is no distributed training happening."
+            )
+
+        # change the default value of `placment_strategy` to `SPREAD`.
+        # using `SPREAD`: the jaxtrainer will spread the workers across different nodes.
+        # between the nodes, the connections is to be established by NCCL;
+        # within the nodes, jax can already enable communicates automatically.
+        # so, no end to use jax backend to make the connections.
+        # NOT `STRICT_SPREAD`: if the users use the trainer on one of GPU nodes instead
+        # of on the extra cpu nodes, then `STRICT_SPREAD` will hang on for the resources
```

Comment:
Hmm not sure if I understand why STRICT_SPREAD won't work. See above comment

Reply:
For example, the placement group will allocate the resources 

```python 
{'cpu': 1}, {'cpu': 1, 'gpu': 1}, {'cpu': 1, 'gpu': 1}, {'cpu': 1, 'gpu': 1}, {'cpu': 1, 'gpu': 1}
```

If the ray cluster only have 4 nodes with one GPU each (i.e. without **CPU** nodes), the resources allocation won't go through?

Reply:
We can set trainer resources to 0, which should avoid this issue